src/i2c.rs

```diff
@@ -0,0 +1,1037 @@
+//! Inter Integrated Circuit (I2C)
+//!
+//! # Terminology
+//! This uses the updated Controller/Target terminology that replaces Master/Slave, as of v7 of the
+//! I2C Spec. See https://www.nxp.com/docs/en/user-guide/UM10204.pdf
+//!
+//! # Usage
+//!
+//! ## Controller
+//!
+//! In the simplest case, the I2C can be initialized from the device peripheral
+//! and GPIO pins:
+//!
+//! ```
+//! let dp = ...;            // Device peripherals
+//! let (scl, sda) = ...;    // GPIO pins
+//!
+//! let mut i2c = dp.I2C1.i2c((scl, sda), 100.kHz(), ccdr.peripheral.I2C1, &ccdr.clocks);
+//! ```
+//!
+//! The GPIO pins should be supplied as a tuple in the following order:
+//!
+//! - Clock (SCL)
+//! - Data (SDA)
+//!
+//! ### Blocking operations
+//!
+//! The driver exposes the controller functionality via the embedded-hal I2C
+//! traits:
+//!
+//! ```
+//! use embedded-hal::i2c::I2c;
+//!
+//! // Write, repeat start, read
+//! let write = [0x11, 0x22, 0x33];
+//! i2c.write(0x18, &write)?;
+//!
+//! // Read only
+//! let mut read = [0u8; 4];
+//! i2c.read(0x18, &mut read)?;
+//!
+//! // Write, repeat start, read
+//! let write = [0x02];
+//! let mut read = [0u8; 4];
+//! i2c.write_read(0x18, &write, &mut read)?;
+//! ```
+//!
+//! ### Non-blocking operation
+//!
+//! The module also exposes non-blocking read/write functions for use with
+//! interrupt-driven or otherwise asynchronous approaches, as well as support
+//! functions to control Start/Stop/Nack operations.
+//!
+//! ```
+//! use nb::block;
+//!
+//! // Simple write
+//! let write = [0x11, 0x22, 0x33];
+//! i2c.start(0x18, AddressMode::AddressMode7bit, write.len(), Stop::Automatic)
+//!
+//! for byte in write {
+//!     block!(i2c.write_nb(byte))?;
```

Comment:
I'd remove the `nb` implementation and instead go for `embedded-hal-async` personally, but up to you